apps/aecore/src/aec_peer_connection.erl

```diff
@@ -613,29 +613,52 @@ get_header(Fun, Arg) ->
 
 %% -- Get Header by Height -----------------------------------------------------
 
-handle_get_header_by_height(S, Msg) ->
+handle_get_header_by_height(S, ?VSN_1, Msg) ->
     Response = get_header(maps:get(height, Msg, undefined)),
     send_response(S, header, Response),
+    S;
+handle_get_header_by_height(S, ?GET_HEADER_BY_HEIGHT_VSN,
+                            #{ height := H, top_hash := TopHash}) ->
+    case {aec_chain:get_header_by_height(H),
+          aec_chain:hash_is_in_main_chain(TopHash)} of
+        {{ok, Header}, true} ->
+            SerHeader = aec_headers:serialize_to_binary(Header),
+            send_response(S, header, {ok, #{ hdr => SerHeader }});
+        {_, false} ->
+            send_response(S, header, {error, not_on_chain});
+        {Err, _} when Err == error orelse Err == {error, chain_too_short} ->
+            send_response(S, header, {error, header_not_found})
+    end,
     S.
 
 %% Response is handled above in Get Header by Hash
 
 %% -- Get N Successors -------------------------------------------------------
 
-handle_get_n_successors(S, Msg) ->
+do_get_n_successors(Hash, N) ->
+    case aec_chain:get_at_most_n_headers_forward_from_hash(Hash, N+1) of
+        {ok, [_ | Headers]} ->
+            HHashes = [ begin
+                            {ok, HHash} = aec_headers:hash_header(H),
+                            <<(aec_headers:height(H)):64, HHash/binary>>
+                        end || H <- Headers ],
+            {ok, #{ header_hashes => HHashes }};
+        error ->
+            {error, block_not_found}
+    end.
+
+handle_get_n_successors(S, Vsn, Msg) ->
     Response =
         case Msg of
-            #{ hash := Hash, n := N } ->
-                 case aec_chain:get_at_most_n_headers_forward_from_hash(Hash, N+1) of
-                     {ok, [_ | Headers]} ->
-                         HHashes = [ begin
-                                         {ok, HHash} = aec_headers:hash_header(H),
-                                         <<(aec_headers:height(H)):64, HHash/binary>>
-                                     end || H <- Headers ],
-                         {ok, #{ header_hashes => HHashes }};
-                     error ->
-                         {error, block_not_found}
-                 end;
+            #{ hash := Hash, n := N } when Vsn == ?VSN_1 ->
+                do_get_n_successors(Hash, N);
+            #{ from_hash := FromHash, target_hash := TargetHash, n := N }
+                when Vsn == ?GET_N_SUCCESSORS_VSN ->
+                Res = do_get_n_successors(FromHash, N),
+                case aec_chain:hash_is_in_main_chain(TargetHash) of
+                    true  -> Res;
+                    false -> {error, not_on_chain}
```

Comment:
I see this as new error code if request is new version. Ok.

Reply:
Not really, this is the error you get if the peer has switched fork.